### PR TITLE
Fixes compile errors in MSVC build #27332

### LINKED ISSUE #27332: Bitcoin fails to build on MSVC: Libevent and BCryptGenRandom errors
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behaviour

When running `msbuild build_msvc\bitcoin.sln -property:Configuration=Release -maxCpuCount -verbosity:minimal` I get the following two errors in the build.

`C:\Users\e0\Documents\GitHub\bitcoin\src\httpserver.cpp(637,9): error C2664: 'void evhttp_connection_get_peer(evhttp_connection *,const char **,uint16_t *)': cannot convert argument 2 from 'char **' to 'const char **' [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\libbitcoin_node\libbitcoin_node
.vcxproj]`

`event.lib(evutil_rand.c.obj) : error LNK2019: unresolved external symbol BCryptGenRandom referenced in function arc4_seed [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\bitcoin-cli\bitcoin-cli.vcxproj]`
C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-cli.exe : fatal error LNK1120: 1 unresolved externals [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\bitcoin-cli\bitcoin-cli.vcxproj]`

Note that vcpkg believes that the latest stable version of libevent is installed

```
PS C:\Users\e0\Documents\GitHub\bitcoin> vcpkg install libevent                                                                        
warning: Starting with the September 2023 release, the default triplet for vcpkg libraries will change from x86-windows to the detected host triplet (x64-windows). To resolve this message, add --triplet x86-windows to keep the same behavior.
Computing installation plan...
The following packages are already installed:
    libevent[core,thread]:x86-windows -> 2.1.12+20230128
libevent:x86-windows is already installed
Restored 0 package(s) from C:\Users\e0\AppData\Local\vcpkg\archives in 98.5 us. Use --debug to see more details.
```

The error message is very similar to this issue https://github.com/bitcoin/bitcoin/issues/23606 which was resolved last year (Jan 13, 2022) with this PR: https://github.com/bitcoin/bitcoin/pull/23607

**A theory:** Given that the PR #23607 solves the issue by redefining `evhttp_connection` in configure.ac. I am speculating the configure.ac is not run for MSVC and thus PR #23607 does not solve this issue for MSCV. Further confirming this is if I add the line `#define HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` to httpserver.cpp this evhttp_connection error no longer appears in the compiler out. However the `BCryptGenRandom` error still occurs.

**Update-1:**
I have solved the bcrypt errors by adding bcrypt as a dependency to the common.init.vcxproj with the following edit:

```xml
    <Link>
      <SubSystem>Console</SubSystem>
      <AdditionalDependencies>bcrypt.lib;Iphlpapi.lib;ws2_32.lib;Shlwapi.lib;kernel32.lib;user32.lib;gdi32.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
      <RandomizedBaseAddress>true</RandomizedBaseAddress>
    </Link>
```

So at minimum this should be added to common.init.vcxproj.in so that `python .\build_msvc\msvc-autogen.py` will add this lib when it builds common.init.vcxproj.

With regards to to the `evhttp_connection` errors, `msvc-autogen.py` does not attempt to define `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` when it reads configure.ac. It might be reasonable to just assume that anyone doing an MSVC build will use vcpk and get the latest version of libevent and thus hardcode `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` in the config for msvc.

**Update-2:** I have created a PR to fix this.

### Expected behaviour

Bitcoin compiles successfully

### Steps to reproduce

I'm following the build instructions here: https://github.com/bitcoin/bitcoin/tree/master/build_msvc

1. Download Bitcoin source code from master branch
2. Install vcpkg and add vcpkg to path so that it actually works,
3. Skip installing QT step and delete QT build targets from bitcoin.sln
4. Run `python build_msvc\msvc-autogen.py`
5. Run `msbuild build_msvc\bitcoin.sln -property:Configuration=Release -maxCpuCount -verbosity:minimal`
6. Build fails with errors shown above

### Relevant log output

```
PS C:\Users\e0\Documents\GitHub\bitcoin> msbuild build_msvc\bitcoin.sln -property:Configuration=Release -maxCpuCount -verbosity:minimal
MSBuild version 17.5.1+f6fdcf537 for .NET Framework

  libunivalue.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libunivalue\libunivalue.lib
  libbitcoin_cli.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_cli\libbitcoin_cli.lib
  libbitcoin_zmq.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_zmq\libbitcoin_zmq.lib
  libsecp256k1.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libsecp256k1\libsecp256k1.lib
  libbitcoin_crypto.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_crypto\libbitcoin_crypto.lib
  libminisketch.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libminisketch\libminisketch.lib
  libbitcoin_wallet_tool.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_wallet_tool\libbitcoin_wallet_tool.lib
  libleveldb.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libleveldb\libleveldb.lib
  libbitcoinconsensus.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoinconsensus\libbitcoinconsensus.lib
  libbitcoin_util.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_util\libbitcoin_util.lib
  libtest_util.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libtest_util\libtest_util.lib
  httpserver.cpp
  libbitcoin_common.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_common\libbitcoin_common.lib
  libbitcoin_wallet.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\libbitcoin_wallet\libbitcoin_wallet.lib
  bitcoin-tx.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-tx.exe
  bitcoin-util.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-util.exe
C:\Users\e0\Documents\GitHub\bitcoin\src\httpserver.cpp(637,9): error C2664: 'void evhttp_connection_get_peer(evhttp_connection *,const char **,uint16_t *)': cannot convert argument 2 from 'char **' to 'const char **' [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\libbitcoin_node\libbitcoin_node
.vcxproj]
C:\Users\e0\Documents\GitHub\bitcoin\src\httpserver.cpp(637,49): message : Conversion loses qualifiers [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\libbitcoin_node\libbitcoin_node.vcxproj]
C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\vcpkg_installed\x64-windows-static\x64-windows-static\include\event2\http.h(1001,6): message : see declaration of 'evhttp_connection_get_peer' [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\libbitcoin_node\libbitcoin_node.vcxproj]
C:\Users\e0\Documents\GitHub\bitcoin\src\httpserver.cpp(637,9): message : while trying to match the argument list '(evhttp_connection *, char **, uint16_t *)' [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\libbitcoin_node\libbitcoin_node.vcxproj]
     Creating library C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-cli.lib and object C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-cli.exp
event.lib(evutil_rand.c.obj) : error LNK2019: unresolved external symbol BCryptGenRandom referenced in function arc4_seed [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\bitcoin-cli\bitcoin-cli.vcxproj]
C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-cli.exe : fatal error LNK1120: 1 unresolved externals [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\bitcoin-cli\bitcoin-cli.vcxproj]
  bitcoin-wallet.vcxproj -> C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-wallet.exe
```

### How did you obtain Bitcoin Core

Compiled from source

### What version of Bitcoin Core are you using?

bitcoin master branch at commit 4c6b7d330a4e80460dcce19b1a0a47d77a0b99ea

### Operating system and version

Windows 11 Pro Version	22H2 Installed on ‎2/‎18/‎2023 OS build	22621.1413 Experience	Windows Feature Experience Pack 1000.22639.1000.0

### Machine specifications

Processor	11th Gen Intel(R) Core(TM) i7-11700F @ 2.50GHz   2.50 GHz
Installed RAM	16.0 GB (15.9 GB usable)
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

### PR DESCRIPTION
**Pull request by EthanHeilman**
This PR is designed to address the issue https://github.com/bitcoin/bitcoin/issues/27332. The MSVC build is failing because of two bugs in how the build is configured.

The issue
====

When running `msbuild build_msvc\bitcoin.sln -property:Configuration=Release -maxCpuCount -verbosity:minima`l the build fails with following two errors.

* `C:\Users\e0\Documents\GitHub\bitcoin\src\httpserver.cpp(637,9): error C2664: 'void evhttp_connection_get_peer(evhttp_connection *,const char **,uint16_t *)': cannot convert argument 2 from 'char **' to 'const char **' [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\libbitcoin_node\libbitcoin_node .vcxproj]`

This error is occurs because bitcoin is using the wrong function signature for `evhttp_connection_get_peer` in libevent. In automake builds, configure.ac inspects the version of libevent it is building against and then defines `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR` to flag the source code to use the correct signature. In MSVC build there does not appear to be a mechanism to do this. So it uses the wrong signature and fails. See the PR https://github.com/bitcoin/bitcoin/pull/23607 for when this logic was added to automake builds.

* `event.lib(evutil_rand.c.obj) : error LNK2019: unresolved external symbol BCryptGenRandom referenced in function arc4_seed [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\bitcoin-cli\bitcoin-cli.vcxproj]
C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\x64\Release\bitcoin-cli.exe : fatal error LNK1120: 1 unresolved externals [C:\Users\e0\Documents\GitHub\bitcoin\build_msvc\bitcoin-cli\bitcoin-cli.vcxproj]`

This error is caused by msbuild not being able to find the library bcrypt.lib because it has not been configured to use bcrypt.lib. 

Fixes (Initially Proposed)
====
**Edit:** This bug was solved using an alternative fix documented below as Fixes (Actual)

While for automake builds a macro is being define to configure the current function signature for `evhttp_connection_get_peer` in libevent, this macro is not being defined for MSVC builds.

1.  This PR addresses this issue by assuming more recent version of libevent is installed and always defining `HAVE_EVHTTP_CONNECTION_GET_PEER_CONST_CHAR`. This logic is more brittle the automake logic, but someone following the MSVC build instructions should only get the more recent version of libevent.

2. This PR fixes the bcrypt.lib errors this by setting this library as a dependency in build_msvc/common.init.vcxproj.in.

Fixes (Actual)
====

* This PR Pin the version of libevent to an older version which uses the older function signature and does not require bcrypt as a separate dependency.

If at some point we have not moved to CMake and we wish upgrade libevent we need to perform the actions in Fixes (Initially Proposed). However it is likely Bitcoin will be on CMake before that is neccessary.

### CODE REVIEW

File: build_msvc/bitcoin_config.h.in [hunk 1]
```diff
@@ -150,6 +150,9 @@
 /* ::wsystem is available */
 #define HAVE_SYSTEM 1
 
+/* Assume that an MSVC build will get dependencies from vcpk and thus use the latest version of libevent */
```
Comment (hernanmarino) [inline build_msvc/bitcoin_config.h.in:153]: Is there a chance that this is not the case in any situation ?

**Review by hernanmarino — COMMENTED** [review]
Concept ACK, but not really sure about the approach (see review comment). Will try to test later this week.

File: build_msvc/bitcoin_config.h.in (hunk 1 above)
Comment (EthanHeilman) [reply build_msvc/bitcoin_config.h.in:153]: I don't know. I would prefer a solution more in line with how this was solved in automake builds but I don't have time to learn enough about msbuild to implement it and it is currently broken in the default case.

Comment (EthanHeilman) [reply build_msvc/bitcoin_config.h.in:153]: In fact the build in CI is now failing because it expects the other function signature.

`C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\httpserver.cpp(635,56): error C2664: 'void evhttp_connection_get_peer(evhttp_connection *,char **,uint16_t *)': cannot convert argument 2 from 'const char **' to 'char **' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\libbitcoin_node\libbitcoin_node.vcxproj]`

Comment (hebasto) [inline build_msvc/bitcoin_config.h.in:153]:
```suggestion
/* Assume that an MSVC build will get dependencies from vcpkg and thus use the latest version of libevent */
```

File: build_msvc/common.init.vcxproj.in [hunk 2]
```diff
@@ -95,7 +95,7 @@
     </ClCompile>
     <Link>
       <SubSystem>Console</SubSystem>
-      <AdditionalDependencies>Iphlpapi.lib;ws2_32.lib;Shlwapi.lib;kernel32.lib;user32.lib;gdi32.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <AdditionalDependencies>bcrypt.lib;Iphlpapi.lib;ws2_32.lib;Shlwapi.lib;kernel32.lib;user32.lib;gdi32.lib;comdlg32.lib;advapi32.lib;shell32.lib;ole32.lib;oleaut32.lib;uuid.lib;odbc32.lib;odbccp32.lib;%(AdditionalDependencies)</AdditionalDependencies>
```
Comment (hebasto) [inline build_msvc/common.init.vcxproj.in:98]: Why this change is required?

**Review by hebasto — COMMENTED** [review]
> In fact the build in CI is now failing because it expects the other function signature.

The recent vcpkg [release](https://github.com/microsoft/vcpkg/releases/tag/2023.02.24) has updated the libevent:
> - libevent `2.1.12#7` -> `2.1.12+20230128#0`

To fix CI [build](https://cirrus-ci.com/task/6327801003900928) we have to:
```diff
--- a/.cirrus.yml
+++ b/.cirrus.yml
@@ -103,7 +103,7 @@ task:
   env:
     PATH: 'C:\jom;C:\Python39;C:\Python39\Scripts;C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\MSBuild\Current\Bin;%PATH%'
     PYTHONUTF8: 1
-    CI_VCPKG_TAG: '2023.01.09'
+    CI_VCPKG_TAG: '2023.02.24'
     VCPKG_DOWNLOADS: 'C:\Users\ContainerAdministrator\AppData\Local\vcpkg\downloads'
     VCPKG_DEFAULT_BINARY_CACHE: 'C:\Users\ContainerAdministrator\AppData\Local\vcpkg\archives'
     CCACHE_DIR: 'C:\Users\ContainerAdministrator\AppData\Local\ccache'
```

For example see https://github.com/bitcoin/bitcoin/commit/1039ed44fb445fb4e58e1990bd07be15b1dbdd15.

File: build_msvc/common.init.vcxproj.in (hunk 2 above)
Comment (EthanHeilman) [reply build_msvc/common.init.vcxproj.in:98]: As far as I can tell the bcrypt.lib dependency was introduced into libevent in 2020. https://github.com/libevent/libevent/commit/138a408c771c21668bcc605c2df00d0384d476d8

**Comment by EthanHeilman** [comment]
@hebasto Thanks for the help for getting the CI/CD fixed. The fact that the version of libevent was being pinned to the earlier version explains why testing didn't catch this.

File: build_msvc/bitcoin_config.h.in (hunk 1 above)
Comment (EthanHeilman) [reply build_msvc/bitcoin_config.h.in]: Using the latest version of libevent in CI is now fixes the build.

**Comment by EthanHeilman** [comment]
@hebasto Thats a great point. I had only been thinking of users that are in the same boat as me, doing a fresh MSVC build, but you are right, most users are likely building against pre-installed packages. I don't think my PR should be merged until it either: 

* Adding the ability to define the function signature based on the actual version of libevent installed
OR
* Pinning the version of libevent in vcpkg.json so that users with old dependencies upgrade when they run build

I'm fairly allergic to the idea of build scripts inspecting function signatures and then defining macros based on that. I understand it makes sense sometimes, but thankfully I don't think we need to do it in this case. I'd advocate for version pinning in vcpkg.json for the follow reasons:

1. Version pinning is a more predictable and straightforward approach.

2. It also enables the ability to force upgrade when dependencies have vulnerabilities.

3. Makes sure everyone is building from the same dependencies which helps identify root causes of bugs. 

4. It lets us ensure that the CI is using the same version of dependencies as someone building from scratch. Likely version pinning would have caught this bug in tests.

When I get free from work tonight I'll add the version to vcpkg.json in the PR.

**Comment by hebasto** [comment]
> I'm fairly allergic to the idea of build scripts inspecting function signatures and then defining macros based on that.

FWIW, the [coming](https://github.com/bitcoin/bitcoin/pull/25797) CMake-based build system will do it :)

**Comment by EthanHeilman** [comment]
@hernanmarino I have libevent pinned to the version with the matching signature function. Let me know if you want anything else changed in this PR.

**Comment by sipsorcery** [comment]
tACK 28436965b6422bd92d23033a5ddbd60a6c183cd7.

The version pinning is nice. It was badly missed feature when we first started relying on vcpkg. We pinned the version of vcpkg instead but individual package pinning should be more flexible.

File: build_msvc/vcpkg.json [hunk 3]
```diff
@@ -13,5 +13,12 @@
...
+  ],
+  "builtin-baseline": "b86c0c35b88e2bf3557ff49dc831689c2f085090",
+  "overrides": [
+    {
+      "name": "libevent",
+      "version": "2.1.12+20230128"
+    }
   ]
-}
+}
```
Comment (hebasto) [inline build_msvc/vcpkg.json:24]: Missed EOL?

File: build_msvc/vcpkg.json [hunk 4]
```diff
@@ -13,5 +13,12 @@
       "features": ["thread"]
     },
     "zeromq"
+  ],
+  "builtin-baseline": "b86c0c35b88e2bf3557ff49dc831689c2f085090",
+  "overrides": [
+    {
+      "name": "libevent",
+      "version": "2.1.12+20230128"
```
[the remark this answers is not in the document]
Comment (EthanHeilman) [reply build_msvc/vcpkg.json:21]: I think that is a fair point. The issue is that if we ever need to upgrade libevent due to security vulnerability it will break and you'll have to figure this out again and change `build_msvc/bitcoin_config.h.in`. This addresses that issue today rather then at some point in the future. That said, if we expect to move to CMake soon, then it makes sense to move to the previous version and get the minimum diff.

I'll change this to `2.1.12#7`, but I'm open to changing to back to `2.1.12+20230128`.

**Comment by EthanHeilman** [comment]
Pushed this on my lunch break. With @hebasto's help this is now a single file, 8 line change.  =)

File: build_msvc/common.init.vcxproj.in (hunk 2 above)
Comment (EthanHeilman) [reply build_msvc/common.init.vcxproj.in]: With `2.1.12#7` we don't need to include `bcrypt.lib` as an additional dependency and so I have removed it. If we upgrade libevent we will need to add this.

File: build_msvc/vcpkg.json [hunk 5]
```diff
@@ -13,5 +13,12 @@
       "features": ["thread"]
     },
     "zeromq"
+  ],
+  "builtin-baseline": "b86c0c35b88e2bf3557ff49dc831689c2f085090",
```
Comment (hebasto) [inline build_msvc/vcpkg.json:17]: This baseline implies package versions as they were at `2022.02.23` tag, which means older ones than they are in our master branch for now. For details, see https://cirrus-ci.com/task/4695704203952128.

Suggesting to update it up to the `2023.01.09` tag:
```suggestion
  "builtin-baseline": "f14984af3738e69f197bf0e647a8dca12de92996",
```

**Review by hebasto — APPROVED** [review]
ACK c40c9edf9266bac0b3553a3d6c5b97f94c34312d

File: build_msvc/vcpkg.json (hunk 5 above)
Comment (EthanHeilman) [reply build_msvc/vcpkg.json:17]: done

**Review by hebasto — APPROVED** [review]
re-ACK 6a9a4d13b2b22632e0acd4f86f7bac238294ba42
